pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -579,6 +582,13 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun, get
 	// Reset the skipped status to trigger recalculation
 	pipelineRunFacts.ResetSkippedCache()
 
+	// If the pipelinerun has timed out, mark tasks as timed out and update status
```

Comment:
I think this is generally the right approach, but my concern is that this will only work for `timeouts.Pipeline` and not `timeouts.Tasks` or `timeouts.Finally`.

Reply:
Ok, so this is all a bit messy and could definitely use a refactoring, but...

`getTaskRunTimeout` calls `calculateTaskRunTimeout` with `timeouts.Tasks` as the `timeout` parameter if `timeout.Tasks` is set, and otherwise with `timeouts.Pipeline`. So that covers `timeouts.Tasks`.

`getFinallyTaskRunTimeout` first calls `calculateTaskRunTimeout` with `timeouts.Pipeline` as the `timeout` parameter. Then it returns the lesser of the result of `calculateTaskRunTimeout` or `timeouts.Finally`. Which, in practice, should be the smallest of `PipelineTask.Timeout`, `timeouts.Finally`, or `timeouts.Pipeline`.

Reply:
fwiw, I'd like to defer a refactor of the various functions/flow/etc for after this PR. I'd prefer to keep this PR focused on the behavioral change.

Reply:
Just to make sure I get it now - what we want is for `timeouts.Tasks` to result in task 1 getting the full value, task 2 to get the value minus the time since the `PipelineRun` started, etc? `timeouts.Finally` is a bit more confusing to me - we don't have a timestamp for when we started processing `finally` tasks, so it's harder to come up an equivalent bit of logic...

Reply:
Yeah exactly. For finally tasks, they all run in parallel, so they can all just get the smaller of (finally timeout, pipeline timeout - elapsed time). The one exception is retries of finally tasks which is a problem we haven't solved yet.

Reply:
Really, in general, I'm not sure what the point of `timeouts.Finally` is? `timeouts.Pipeline` is obvious, and I get `timeouts.Tasks` is meant to ensure that the non-finally tasks only take up so much of the run, but even before this change, what was the practical impact of `timeouts.Finally`? 

If `timeouts.Tasks+timeouts.Finally < timeouts.Pipeline`, then we never hit `timeouts.Pipeline`, and if `timeouts.Tasks+timeouts.Finally == timeouts.Pipeline`, then we're always timing out finally tasks based on the overall `PipelineRun` timeout anyway...

Reply:
I think you can specify `timeouts.Finally` without specifying `timeouts.Pipeline`

Reply:
Actually, in practice, if you don't specify `timeouts.Pipeline`, you get a default timeout of 60m for the whole `PipelineRun`, and if you specify 0 for `timeouts.Pipeline`, you'll get a validation error. I think. It's all a bit messy. It feels like if we want to support the scenario of no timeout on the `PipelineRun` as a whole but an explicit timeout just for finally tasks, we'd need to pass validation when `timeouts.Finally` is greater than `timeouts.Pipeline` and `timeouts.Pipeline` is 0 (aka `config.NoTimeoutDuration`), and then add a field to `PipelineRunStatusFields` like `FinallyStartTime` to track when we started running the finally tasks...

Reply:
I'm seriously tempted to add new `TasksStartTime` and `FinallyStartTime` fields to `PipelineRunStatusFields`, and rather than setting explicit timeouts on child `TaskRun`s and `Run`s, have the `PipelineRun` reconciler check `now - TasksStartTime > Timeouts.Tasks` and `now - FinallyStartTime > Timeouts.Finally`, then cancel/timeout the child tasks when that's the case...

Reply:
(though when I took a quick look to see how one would go about adding `FinallyStartTime`, I didn't find an obvious way to tell when we've switched to executing finally tasks)

Reply:
I think this conversation can be resolved!